docs/decisions/0006-general-principles-for-react-development.md

```diff
@@ -0,0 +1,29 @@
+# General Principles for React development
+
+## Motivation
+
+Certain decisions in these ADRs are based on the goal to make our code better. But what does "better" mean? Let's define it by creating some principles for good code quality when developing React in this project.
+
+## Principles
+
+I. We aim to reduce complexity in order to speed up development, reduce risk of bugs, and make it easy to change code.
+
+Reason: This is self-explanatory, however not always followed in practice. We are developing this project now, and complexity needs to be tackled from the start, before it grows too much. The alternative is to ending with a project that is extremely hard to understand and work on and resistent to change, like the current edx-platform. Resistance to change is what slows developer teams down.
+
+II. To achieve the above point, we generally orient ourselves at Clean Code principles.
+
+Reason: Clean Code is industry standard and proven method to tackle complexity.
+
+III. We follow the React docs and prefer patterns and principles that are established in the community in order to have a good architecture that is easy to understand.
+
+Reason: The community has spent years learning and improving upon what works well and what doesn't, and this has led to the patterns described there. The docs are in general not opinionated and violating their principles has caused applications to have many bugs. When we use established principles, we make it easy for people to understand the code, especially if they are newcomers to the project or community members.
+
+IV. We follow the KISS principle - "Keep it simple, stupid", in order to make it easy for any developer to understand the code.
+
+Reason: We work agile and with an open source community. A common rule of thumb is that a developer with just 3 months of experience should be able to understand code. We should enable developers to work on this codebase without spending a lot of time trying to understand what is going on.
+
+V. We do not increase code complexity in order to make it easier to write tests, unless absolutely necessary and for good reasons. (Reducing complexity to improve testability is fine.)
+
```

Comment:
Yes and no. Using dependency injection to make it easy to mock out dependencies is essential for testability. One might argue that the inversion of control that comes with that is non-intuitive and makes the code harder to understand. But it's a really sound design and testability principle. We're in agreement if dependency injection is not viewed as increasing code complexity.

Reply:
You mentioned the KISS design principle above, but we should equally be adhering to other design principles such as the single-responsibility principle. For example, mixing UI and business logic may be intuitive when first implementing a feature, but may well make the testing harder and may actually be harder to use in the future, as requirements change. This is probably addressed in the open source community practices that you've referenced, but I worry about a blanket statement that we shouldn't structure our code for testability.

Put another way, following the best design principles is good engineering and is good for testability, but it may not be ideal for clarity to someone with three months' experience. I believe a more appropriate metric is that the code is easily accessible to someone with limited domain knowledge, but who has experience following good design principles.

Reply:
I fully agree with these points. It would be great to find a better wording for this. My motivation for the above statement V. is that I see some unusual patterns in our codebase that I haven't seen used much in React codebases, which I consider bad / unnecessarily complex patterns, and they have been explicitly justified in an earlier ADR with saying that improved testability should justify higher complexity. I am attempting to understand the underlying problem, and it seems to me to have something to do with believing there is a tradeoff - that for higher testability you need to increase code complexity. My experience is usually that higher testability best goes hand in hand with a really good code design. I would also add to your first point that dependency injection as you say is "a really sound design" principle as well, which justifies it. We don't really have the option to do this easily in React, but I want to make the point that we should improve testability **along with** good design and not **instead of** good design.
Do you have any ideas how to frame this point in a way that is better and less of a blanket statement?